### PR TITLE
remove all nodes from queue on the activation of staking v4.

### PR DESCRIPTION
**Pull request by sasurobert**
## Reasoning behind the pull request
- unstake all nodes from the queue when activating staking v4
  
## Proposed changes
- 
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: vm/systemSmartContracts/stakingWaitingList.go [hunk 1]
```diff
@@ -801,6 +801,60 @@ func (s *stakingSC) stakeNodesFromQueue(args *vmcommon.ContractCallInput) vmcomm
...
+
+	nodePriceToUse := big.NewInt(0).Set(s.minNodePrice)
+	if s.enableEpochsHandler.IsFlagEnabled(common.CorrectLastUnJailedFlag) {
+		nodePriceToUse.Set(s.stakeValue)
+	}
+
+	for i, blsKey := range waitingListData.blsKeys {
+		registrationData := waitingListData.stakedDataList[i]
+
+		registrationData.Staked = false
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/stakingWaitingList.go:836]: should we reuse the function tryUnStake for L836-L845?

File: epochStart/metachain/systemSCs.go [hunk 2]
```diff
@@ -139,7 +138,7 @@ func (s *systemSCProcessor) processWithNewFlags(
 	}
 
 	if s.enableEpochsHandler.IsFlagEnabled(common.StakingV4Step1Flag) {
-		err := s.stakeNodesFromQueue(validatorsInfoMap, math.MaxUint32, header.GetNonce(), common.AuctionList)
+		err := s.unStakeAllNodesFromQueue()
```
Comment (iulianpascalau) [inline epochStart/metachain/systemSCs.go:141]: does the `stakeNodesFromQueue` is still used?

File: epochStart/metachain/systemSCs.go [hunk 3]
```diff
@@ -170,6 +169,32 @@ func (s *systemSCProcessor) processWithNewFlags(
...
+	}
+	vmOutput, errRun := s.systemVM.RunSmartContractCall(vmInput)
+	if errRun != nil {
+		return fmt.Errorf("%w when unStaking all nodes from waiting list", errRun)
+	}
+	if vmOutput.ReturnCode != vmcommon.Ok {
+		return fmt.Errorf("got return code %s when unStaking all nodes from waiting list", vmOutput.ReturnCode)
+	}
+
+	err := s.processSCOutputAccounts(vmOutput)
```
Comment (iulianpascalau) [inline epochStart/metachain/systemSCs.go:190]: return directly?

File: epochStart/metachain/systemSCs.go (hunk 2 above)
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:141]: in old code yes.

File: vm/systemSmartContracts/stakingWaitingList.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/stakingWaitingList.go:836]: we cannot use that one, as that is for staked nodes, not nodes in queue.

File: vm/systemSmartContracts/stakingWaitingList.go [hunk 4]
```diff
@@ -801,6 +801,60 @@ func (s *stakingSC) stakeNodesFromQueue(args *vmcommon.ContractCallInput) vmcomm
...
+	if err != nil {
+		s.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if len(waitingListData.blsKeys) == 0 {
+		s.eei.AddReturnMessage("no nodes in queue")
+		return vmcommon.Ok
+	}
+
+	nodePriceToUse := big.NewInt(0).Set(s.minNodePrice)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/stakingWaitingList.go:828]: this is not needed.

File: epochStart/metachain/systemSCs.go [hunk 5]
```diff
@@ -170,6 +169,32 @@ func (s *systemSCProcessor) processWithNewFlags(
...
+			CallerAddr: vm.EndOfEpochAddress,
+			CallValue:  big.NewInt(0),
+			Arguments:  [][]byte{},
+		},
+		RecipientAddr: vm.StakingSCAddress,
+		Function:      "unStakeAllNodesFromQueue",
+	}
+	vmOutput, errRun := s.systemVM.RunSmartContractCall(vmInput)
+	if errRun != nil {
+		return fmt.Errorf("%w when unStaking all nodes from waiting list", errRun)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:184]: ok

File: vm/systemSmartContracts/staking_test.go [hunk 6]
```diff
@@ -3591,3 +3591,116 @@ func TestStakingSc_fixMissingNodeAddAsLast(t *testing.T) {
...
+	arguments.Function = "stakeNodesFromQueue"
+	retCode = stakingSmartContract.Execute(arguments)
+	assert.Equal(t, retCode, vmcommon.Ok)
+
+	for i := currentOutPutIndex; i < len(eei.output); i += 2 {
+		checkIsStaked(t, stakingSmartContract, arguments.CallerAddr, eei.output[i], vmcommon.Ok)
+	}
+	assert.Equal(t, 6, len(eei.output)-currentOutPutIndex)
+	stakingConfig := stakingSmartContract.getConfig()
+	assert.Equal(t, stakingConfig.StakedNodes, int64(4))
```
Comment (mariusmihaic) [inline vm/systemSmartContracts/staking_test.go:3705]: We can also a check here

```go
	waitingListHead, _ := stakingSmartContract.getWaitingListHead()
	require.Equal(t, waitingListHead.Length, 0)
```

**Review by mariusmihaic — APPROVED** [review]

File: vm/systemSmartContracts/staking_test.go (hunk 6 above)
Comment (iulianpascalau) [reply vm/systemSmartContracts/staking_test.go:3705]: refactored the test

**Review by AdoAdoAdo — APPROVED** [review]

File: integrationTests/vm/staking/stakingV4_test.go [hunk 7]
```diff
@@ -933,9 +946,11 @@ func TestStakingV4_JailAndUnJailNodes(t *testing.T) {
 	newJailed := getAllPubKeys(currNodesConfig.waiting)[:1]
 	node.ProcessJail(t, newJailed)
 
+	// TODO fix the test below this point
```
Comment (iulianpascalau) [inline integrationTests/vm/staking/stakingV4_test.go:949]: Need help with this test

**Review by iulianpascalau — COMMENTED** [review]
:+1:

File: integrationTests/vm/staking/stakingV4_test.go [hunk 8]
```diff
@@ -216,8 +216,7 @@ func TestStakingV4(t *testing.T) {
 	require.Len(t, getAllPubKeys(nodesConfigStakingV4Step1.waiting), totalWaiting)
 	require.Empty(t, nodesConfigStakingV4Step1.queue)
 	require.Empty(t, nodesConfigStakingV4Step1.shuffledOut)
-	// the queue should be empty
-	requireSameSliceDifferentOrder(t, make([][]byte, 0), nodesConfigStakingV4Step1.auction)
+	require.Empty(t, nodesConfigStakingV4Step1.auction) // the queue should be empty
```
Comment (iulianpascalau) [inline integrationTests/vm/staking/stakingV4_test.go:219]: :+1:

**Review by iulianpascalau — COMMENTED** [review]
:+1:

**Review by AdoAdoAdo — DISMISSED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
Normal allin test: v1.6.15-dev-config-eb2e06c06d -> unStake-from-queue-at-acti-9cf69bdb91

--- Specific errors ---

block hash does not match 228
wrong nonce in block 192
miniblocks does not match 0
num miniblocks does not match 0
miniblock hash does not match 0
block bodies does not match 0
receipts hash missmatch 0

/------/


--- Statistics ---

Nr. of all ERRORS:  0
Nr. of all WARNS:  260
Nr. of new ERRORS:  0
Nr. of new WARNS:  1
Nr. of PANICS:  0

/------/

--- ERRORS ---

/------/

--- WARNINGS ---

/------/

**Review by ssd04 — APPROVED** [review]

**Pull request merged**